File: demos/Simple Demo/Simple Demo/SimpleTableViewController.m
//
//  StoryboardExampleTableViewController.m
//  Simple Demo
//
//  Created by Muthu Arumugam on 10/6/15.
//  Copyright © 2015 Yieldmo, Inc. All rights reserved.
//

#import "SimpleTableViewController.h"
@import Yieldmo;

static NSUInteger const kAdIndex = 3;

@interface SimpleTableViewController () <YMPlacementViewDelegate>

@property (nonatomic, strong) YMPlacementView * adView;
@property (nonatomic, strong) NSString * placementID;

@end

@implementation SimpleTableViewController

- (void) viewDidLoad {
    [super viewDidLoad];

    // Init YM Views
    self.placementID = kDemoPlacement_Domino_4Ads;
    self.adView = [Yieldmo placementWithFrame:CGRectMake(0, 0, self.view.frame.size.width, 600) withID:self.placementID];
    self.adView.delegate = self;
}

- (void) didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger) numberOfSectionsInTableView: (UITableView *) tableView {
    return 1;
}

- (NSInteger) tableView: (UITableView *) tableView numberOfRowsInSection: (NSInteger) section {
    return 20;
}

- (UITableViewCell *) tableView: (UITableView *) tableView cellForRowAtIndexPath: (NSIndexPath *) indexPath {
    UITableViewCell * cell;

    // Configure the cell...
    if (indexPath.row == kAdIndex) {
        // For Yieldmo Ad Cell
        if (self.adView) {
            // If ad request succeeds, fill the row with ad content
            cell = [tableView dequeueReusableCellWithIdentifier: @"placementCellId" forIndexPath: indexPath];
            if (![cell.subviews containsObject: self.adView]) {
                [cell addSubview: self.adView];
            }
        }
        else {
            // Load a dummy row if the ad request fails
            cell = [tableView dequeueReusableCellWithIdentifier: @"cell" forIndexPath: indexPath];
            cell.textLabel.text = @"";
        }
    }
    else {
        cell = [tableView dequeueReusableCellWithIdentifier: @"cell" forIndexPath: indexPath];
        cell.textLabel.text = [NSString stringWithFormat: @"Row %ld", (long)indexPath.row];
    }

    return cell;
}

- (CGFloat) tableView: (UITableView *) tableView heightForRowAtIndexPath: (NSIndexPath *) indexPath {
    if (indexPath.row == kAdIndex) {
        // For Yieldmo Ad Cell
        if (self.adView) {
            return self.adView.frame.size.height; // Ad available
        }
        else {
            return 0.0; // No Ad
        }
    }
    else {
        return 40.0f;
    }
}

#pragma mark - YMPlacementViewDelegate methods

// Optional
- (void) adViewDidReceiveAd: (YMPlacementView *) placementView {
    NSLog(@"Received ad from server for placement: %@", self.placementID);
}

// Optional
- (void) adViewDidFailToReceiveAd: (YMPlacementView *) placementView WithError: (YMError) error {
    NSLog(@"Failed to get ad from server for placement: %@", self.placementID);
}

// Required delegate method
- (BOOL) adViewShouldResize: (YMPlacementView *) placementView toPosition: (CGRect) position {
    return YES;
}

// Optional method that gets called if adViewShouldResize returns YES. Use it to trigger any events that need to happen after the adView resizes
- (void) adViewDidResize: (YMPlacementView *) placementView toPosition: (CGRect) position {
    [self.tableView reloadRowsAtIndexPaths: [NSArray arrayWithObjects: [NSIndexPath indexPathForRow: kAdIndex inSection: 0], nil] withRowAnimation: UITableViewRowAnimationNone];
}

@end

File: demos/Swift/Simple Demo/Simple Demo/SimpleTableViewController.swift
//
//  SimpleTableViewController.swift
//  Simple Demo
//
//  Created by Muthu Arumugam on 10/7/15.
//  Copyright © 2015 Yieldmo, Inc. All rights reserved.
//

import UIKit

class SimpleTableViewController: UITableViewController, YMViewDelegate {
    
    var adContent = YMView()

    override func viewDidLoad() {
        super.viewDidLoad()

        YM.sharedYM().fetchNativePlacements([YMNativePlacement.init(placementId: kDemoDominoFormat4AdPlacementId, width: self.view.bounds.size.width)], delegate: self)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 20
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var cell:UITableViewCell
        
        if  indexPath.row == Int(parallaxRowIndex()) {
            // For Yieldmo Ad Cell
            if (adContent is NilLiteralConvertible) {                
                cell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath)
                cell.textLabel?.text = ""
            } else {
                cell = tableView.dequeueReusableCellWithIdentifier("placementCellId", forIndexPath: indexPath)
                cell.addSubview(adContent)
            }

        } else {
            cell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath)
            cell.textLabel?.text = "Row \(indexPath.row)"
        }
        
        return cell
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        if  indexPath.row == Int(parallaxRowIndex()) {
            if (adContent is NilLiteralConvertible) {
                return 0.0; // Ad available
            } else {
                return self.adContent.preferredHeight; // No Ad
            }
        } else {
            return 40.0;
        }
    }

    // MARK: - YMViewDelegate methods
    
    func placementFailedToFetchAds(erroredPlacements: [AnyObject]!) {
        print("Failed to get ad from server for placements: \(erroredPlacements)")
        tableView!.reloadData()
    }
    
    func placementDidFetchAds(placementViews: [NSObject : AnyObject]!) {
        print("Successfully received an ad for \(placementViews.count) view(s)")
        adContent = placementViews[kDemoDominoFormat4AdPlacementId]! as! YMView
        tableView.reloadData()
    }
    
    override func scrollViewDidScroll(scrollView: UIScrollView) {
        adContent.didScroll()
    }
    
    func placementScrollView() -> UIScrollView! {
        return tableView;
    }
    
    func parallaxRowIndex() -> UInt {
        return 3;
    }
    
    // MARK: Prepare for Wrapper Demo
    @IBAction func wrapperButtonClicked(sender: AnyObject) {
        let wrapperPlacement = YMNativePlacement.init(placementId: kDemoWrapperFormatPlacementId, width: self.view.bounds.size.width)
        let storyboard = UIStoryboard.init(name: "Main", bundle: NSBundle.mainBundle())
        let wrapperViewController = storyboard.instantiateViewControllerWithIdentifier("wrapperdemo")
        let ymVC = YMWrapperViewController.init(childViewController: wrapperViewController, placements: [wrapperPlacement])
        self.navigationController?.pushViewController(ymVC, animated: true)
    }
    

}
